File: c/stack/min-stack.cpp
class MinStack {
public:
    /** initialize your data structure here. */
    MinStack() {

    }

    void push(int x) {
        st.push(x);
        if(min_st.empty() || x <= min_st.top())
            min_st.push(x);
    }

    void pop() {
        int x = st.top();
        if(x == min_st.top())
            min_st.pop();
        st.pop();
    }

    int top() {
        return st.top();
    }

    int getMin() {
        return min_st.top();
    }

private:
    stack<int> st;
    stack<int> min_st;
};

/**
 * Your MinStack object will be instantiated and called as such:
 * MinStack obj = new MinStack();
 * obj.push(x);
 * obj.pop();
 * int param_3 = obj.top();
 * int param_4 = obj.getMin();
 */

File: python/stack/min-stack.py
class MinStack(object):

    def __init__(self):
        self.st = []
        self.min_st = []


    def push(self, x):
        self.st.append(x)
        if len(self.min_st) == 0 or x <= self.min_st[-1]:
            self.min_st.append(x)


    def pop(self):
        if self.st[-1] <= self.min_st[-1]:
            self.min_st.pop()
        return self.st.pop()


    def top(self):
        return self.st[-1]


    def getMin(self):
        return self.min_st[-1]



# Your MinStack object will be instantiated and called as such:
# obj = MinStack()
# obj.push(x)
# obj.pop()
# param_3 = obj.top()
# param_4 = obj.getMin()
